src/main/java/games/dominion/actions/DominionAttackAction.java

```diff
@@ -0,0 +1,138 @@
+    }
+
+    int currentTarget;
+    boolean[] reactionsInitiated;
+    boolean[] attacksInitiated;
+    boolean[] attacksComplete;
+
+    /**
+     * This must be called by the implementing sub-class at the correct point in the card logic
+     *
+     * @param state
+     * @return
+     */
+    void initiateAttack(DominionGameState state) {
+        // for any state changes due to the immediate effects of the card on the player.
+        // then what we need to do is the cycling through each of the other players to allow them
+        // to play Reaction cards, and then suffer the direct attack effects of the card.
+        state.setActionInProgress(this);
+        currentTarget = (player + 1) % state.getNPlayers();
+        reactionsInitiated = new boolean[state.getNPlayers()];
+        attacksComplete = new boolean[state.getNPlayers()];
+        attacksInitiated = new boolean[state.getNPlayers()];
+        reactionsInitiated[player] = true;
+        attacksComplete[player] = true;
+        nextPhaseOfReactionAttackCycle(state);
+    }
+
+    private void nextPhaseOfReactionAttackCycle(DominionGameState state) {
+        // we cycle through each other player in turn to give them the opportunity to play Reactions
+        // Then, after they have played their Reactions, we execute the actual attack code (in the sub-class)
+        // if they are undefended
+        int loopCount = 0;
+        do {
+            if (!(reactionsInitiated[currentTarget] && attacksComplete[currentTarget])) {
+                // at least one of these two booleans must therefore be false
+                if (!reactionsInitiated[currentTarget]) { // we are in AttackReaction phase
+                    AttackReaction reaction = new AttackReaction(state, player, currentTarget);
+                    reactionsInitiated[currentTarget] = true;
+                    if (!reaction.executionComplete(state)) {
+                        // there is some reaction to process - so put that on the stack
+                        state.setActionInProgress(reaction);
+                        return;
+                    }
+                }
+                // we are in now Attack phase phase
+                if (state.isDefended(currentTarget)) {
+                    attacksInitiated[currentTarget] = true;
+                    attacksComplete[currentTarget] = true;
+                } else {
+                    if (!attacksInitiated[currentTarget])
+                        executeAttack(state);
+                    attacksInitiated[currentTarget] = true;
+                    if (isAttackComplete(currentTarget, state)) {
+                        attacksComplete[currentTarget] = true;
+                    } else {
+                        return;
+                    }
+                }
+            }
+            currentTarget = (currentTarget + 1) % state.getNPlayers();
+            loopCount++;
+            if (loopCount > 100) {
+                throw new AssertionError("WTF?");
```

Comment:
Is this to avoid infinite loops? :)

Reply:
Yes. It's purely safety code - initially when I was trying to track down a mysterious infinite loop while debugging. As it turns out, this wasn't the culprit, but I felt leaving this in does no harm (and may do some good).